flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksDBFullSnapshotResources.java

```diff
@@ -0,0 +1,206 @@
+    private final Snapshot snapshot;
+    private final RocksDB db;
+    private final List<MetaData> metaData;
+
+    /** Number of bytes in the key-group prefix. */
+    @Nonnegative private final int keyGroupPrefixBytes;
+
+    private final KeyGroupRange keyGroupRange;
+    private final TypeSerializer<K> keySerializer;
+    private final StreamCompressionDecorator streamCompressionDecorator;
+
+    public RocksDBFullSnapshotResources(
+            ResourceGuard.Lease lease,
+            Snapshot snapshot,
+            List<RocksDBKeyedStateBackend.RocksDbKvStateInfo> metaDataCopy,
+            List<StateMetaInfoSnapshot> stateMetaInfoSnapshots,
+            RocksDB db,
+            int keyGroupPrefixBytes,
+            KeyGroupRange keyGroupRange,
+            TypeSerializer<K> keySerializer,
+            StreamCompressionDecorator streamCompressionDecorator) {
+        this.lease = lease;
+        this.snapshot = snapshot;
+        this.stateMetaInfoSnapshots = stateMetaInfoSnapshots;
+        this.db = db;
+        this.keyGroupPrefixBytes = keyGroupPrefixBytes;
+        this.keyGroupRange = keyGroupRange;
+        this.keySerializer = keySerializer;
+        this.streamCompressionDecorator = streamCompressionDecorator;
+
+        // we need to to this in the constructor, i.e. in the synchronous part of the snapshot
+        // TODO: better yet, we can do it outside the constructor
+        this.metaData = fillMetaData(metaDataCopy);
+    }
+
+    private List<MetaData> fillMetaData(
+            List<RocksDBKeyedStateBackend.RocksDbKvStateInfo> metaDataCopy) {
+        List<MetaData> metaData = new ArrayList<>(metaDataCopy.size());
+        for (RocksDBKeyedStateBackend.RocksDbKvStateInfo rocksDbKvStateInfo : metaDataCopy) {
+            StateSnapshotTransformer<byte[]> stateSnapshotTransformer = null;
+            if (rocksDbKvStateInfo.metaInfo instanceof RegisteredKeyValueStateBackendMetaInfo) {
+                stateSnapshotTransformer =
+                        ((RegisteredKeyValueStateBackendMetaInfo<?, ?>) rocksDbKvStateInfo.metaInfo)
+                                .getStateSnapshotTransformFactory()
+                                .createForSerializedState()
+                                .orElse(null);
+            }
+            metaData.add(new MetaData(rocksDbKvStateInfo, stateSnapshotTransformer));
+        }
+        return metaData;
+    }
+
+    @Override
+    public KeyValueStateIterator createKVStateIterator() throws IOException {
+        CloseableRegistry closeableRegistry = new CloseableRegistry();
+
+        try {
+            ReadOptions readOptions = new ReadOptions();
+            closeableRegistry.registerCloseable(readOptions::close);
+            readOptions.setSnapshot(snapshot);
+
+            List<Tuple2<RocksIteratorWrapper, Integer>> kvStateIterators =
+                    createKVStateIterators(readOptions);
+
+            for (Tuple2<RocksIteratorWrapper, Integer> iter : kvStateIterators) {
```

Comment:
nit: We could do that in a single pass when creating iterators.

Reply:
fixing